python/ql/lib/semmle/python/frameworks/Django.qll

```diff
@@ -645,6 +805,166 @@ module PrivateDjango {
+                c.(DataFlow::AttributeContent).getAttribute() = fieldName and
+                (
+                  // -> object created
+                  (
+                    methodName = "create" and nodeTo = call
+                    or
+                    // TODO: for these two methods, the result is a tuple `(<Model>, bool)`,
+                    // which we need flow-summaries to model properly
+                    methodName in ["get_or_create", "update_or_create"] and none()
+                  )
+                  or
+                  // -> DB store on synthetic node
+                  nodeTo = nodeToStoreIn(modelClass, fieldName)
+                )
+              )
+              or
+              // attribute store in `<Model>.objects.[<QuerySet>].update()` -> synthetic
+              // see https://docs.djangoproject.com/en/4.0/ref/models/querysets/#update
+              exists(DataFlow::CallCfgNode call, API::Node modelClass, string fieldName |
+                call = [manager(modelClass), querySet(modelClass)].getMember("update").getACall() and
+                nodeFrom = call.getArgByName(fieldName) and
+                c.(DataFlow::AttributeContent).getAttribute() = fieldName and
+                nodeTo = nodeToStoreIn(modelClass, fieldName)
+              )
+              or
+              // synthetic -> method-call that returns collection of ORM models (all/filter/...)
+              exists(API::Node modelClass |
+                nodeFrom = nodeToLoadFrom(modelClass) and
+                nodeTo.(Model::QuerySetMethodInstanceCollection).getModelClass() = modelClass and
+                nodeTo.(Model::QuerySetMethodInstanceCollection).isDbFetch() and
+                c instanceof DataFlow::ListElementContent
+              )
+              or
+              // synthetic -> method-call that returns dictionary with ORM models as values
+              exists(API::Node modelClass |
+                nodeFrom = nodeToLoadFrom(modelClass) and
+                nodeTo.(Model::QuerySetMethodInstanceDictValue).getModelClass() = modelClass and
+                nodeTo.(Model::QuerySetMethodInstanceDictValue).isDbFetch() and
+                c instanceof DataFlow::DictionaryElementAnyContent
+              )
+            }
+
+            override predicate jumpStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+              // save -> synthetic
+              exists(API::Node modelClass, DataFlow::MethodCallNode saveCall |
+                // TODO: The `nodeTo` should be restricted more, such that flow to
+                // base-classes are only for the fields that are defined in the
+                // base-class... but only passing on flow for a specific attribute requires flow-summaries,
+                // so we can do
+                // `obj (in obj.save call) ==read of attr==> synthetic attr on base-class ==store of attr==> synthetic for base-class`
+                nodeTo = nodeToStoreIn(modelClass, _) and
```

Comment:
I believe that `nodeToStoreIn(modelClass, _)` does indeed restrict to fields defined by the base class?

Reply:
Not quite. The problem is https://github.com/github/codeql/pull/8061/files/27d5349a745afb6da1fb516276b838ee75afddd5#diff-4899b0abafadf88280bc9f62bb280ebc9c4369071b008e853464b0b184539b3eR162

Since there had been a store of a different subclass with the same field-name

```
class PolyBook(PolymorphicModel):
    title = models.CharField(max_length=256)


class PolyPhysicalBook(PolyBook):
    physical_location = models.CharField(max_length=256)
    same_name_different_value = models.CharField(max_length=256)


class PolyEBook(PolyBook):
    download_link = models.CharField(max_length=256)
    same_name_different_value = models.CharField(max_length=256)

def poly_save_physical_book():
    return PolyPhysicalBook.objects.create(
        title=SOURCE,
        physical_location=SOURCE,
        same_name_different_value=SOURCE,
    )
```

we think there is flow from the SOURCE to the `same_name_different_value` field _on the base class `PolyBook`_ (although that field doesn't exist on the base-class), and therefore also to the subclass `PolyEBook` -- but there is no such flow.